pam/gdm/conversation_test.go

```diff
@@ -99,3 +129,569 @@ func TestSendToGdm(t *testing.T) {
 		})
 	}
 }
+
+func TestDataSend(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(pam_test.MaybeDoLeakCheck)
+	testCases := map[string]struct {
+		value *Data
+
+		wantReturn []byte
+		wantError  error
+	}{
+		"nil return": {
+			value: &Data{
+				Type: DataType_event,
+				Event: &EventData{
+					Type: EventType_brokerSelected,
+					Data: &EventData_BrokerSelected{},
+				},
+			},
+
+			wantReturn: []byte("null"),
+		},
+
+		"version data": {
+			value: &Data{
+				Type:  DataType_hello,
+				Hello: &HelloData{Version: 12345},
+			},
+			wantReturn: []byte(`"hello gdm!"`),
+		},
+
+		// Error cases
+		"Error on empty data": {
+			value:      &Data{},
+			wantReturn: nil,
+			wantError:  errors.New("expected type unknownType"),
+		},
+
+		"Error on missing data return": {
+			value: &Data{
+				Type: DataType_event,
+				Event: &EventData{
+					Type: EventType_brokerSelected,
+					Data: nil,
+				},
+			},
+
+			wantError: errors.New("missing event data"),
+		},
+
+		"Error on wrong data": {
+			value: &Data{
+				Type:    DataType_event,
+				Request: &RequestData{},
+			},
+			wantError: errors.New("missing event type"),
+		},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+			convFuncCalled := true
+			mtx := pam_test.NewModuleTransactionDummy(pam.BinaryPointerConversationFunc(
+				func(ptr pam.BinaryPointer) (pam.BinaryPointer, error) {
+					convFuncCalled = true
+					require.NotNil(t, ptr)
+					req, err := decodeJSONProtoMessage(ptr)
+					require.NoError(t, err)
+					valueJSON, err := tc.value.JSON()
+					require.NoError(t, err)
+					require.Equal(t, valueJSON, req)
+					if tc.wantReturn != nil {
+						msg, err := newJSONProtoMessage(tc.wantReturn)
+						require.NoError(t, err)
+						return pam.BinaryPointer(msg), nil
+					}
+					msg, err := newJSONProtoMessage(req)
+					require.NoError(t, err)
+					return pam.BinaryPointer(msg), tc.wantError
+				}))
+
+			data, err := tc.value.Send(mtx)
+			require.True(t, convFuncCalled)
+
+			if tc.wantError != nil {
+				require.Nil(t, data)
+				require.Error(t, tc.wantError, err)
+				return
+			}
+
+			require.NoError(t, err)
```

Comment:
These are part of the same block for checking the error value, so there should not be a space between the end of the if and the `require.NoError`.